File: getting-started-mocks/02-first-useful-mock/node/notifier/carrier.js
// The production getShipmentStatus: it asks the carrier's HTTP API. This is
// the same code that used to live inside notify(), moved behind the seam.
// Nothing in this post's tests runs it; that is deliberate and the post says why.
export function httpShipmentStatus(baseUrl) {
  return async (trackingNumber) => {
    const response = await fetch(`${baseUrl}/shipments/${trackingNumber}`)
    if (!response.ok) {
      throw new Error(`carrier returned ${response.status}`)
    }
    const shipment = await response.json()
    return shipment.status
  }
}

File: getting-started-mocks/02-first-useful-mock/python/notifier/carrier.py
"""The production ShipmentStatus: it asks the carrier's HTTP API.

This is the same code that used to live inside notify(), moved behind the
seam. Nothing in this post's tests runs it; that is deliberate and the post
says why.
"""

import json
from urllib.request import urlopen

from notifier.notifier import ShipmentStatus


def http_shipment_status(base_url: str) -> ShipmentStatus:
    def lookup(tracking_number: str) -> str:
        with urlopen(f"{base_url}/shipments/{tracking_number}") as response:
            if response.status != 200:
                raise RuntimeError(f"carrier returned {response.status}")
            shipment = json.load(response)
        return shipment["status"]

    return lookup
